tests/testhelpers.py

```diff
@@ -1,257 +1,200 @@
+  keys_obs = list(target_obs.keys())
+  if keys_src != keys_obs:
+    if debug:
+      logging.error("entities don't match")
+    return False
+
+  for k in keys_src:
+    ent_src = source_obs[k]
+    ent_tgt = target_obs[k]
+    if list(ent_src.keys()) != list(ent_tgt.keys()):
+      if debug:
+        logging.error("entries don't match. key: %s", str(k))
+      return False
+
+    obj = ent_src.keys()
+    for o in obj:
+      obj_src = ent_src[o]
+      obj_tgt = ent_tgt[o]
+      if np.sum(obj_src != obj_tgt) > 0:
+        if debug:
+          logging.error("objects don't match. key: %s, obj: %s", str(k), str(o))
+        return False
+
+  return True
+
+
+def player_total(env):
+  return sum(ent.gold.val for ent in env.realm.players.values())
+
+
+def count_actions(tick, actions):
+  cnt_action = {}
+  for atn in (Move, Attack, Sell, Use, Give, Buy):
+    cnt_action[atn] = 0
+
+  for ent_id in actions:
+    for atn, _ in actions[ent_id].items():
+      if atn in cnt_action:
+        cnt_action[atn] += 1
+      else:
+        cnt_action[atn] = 1
+
+  info_str = f"Tick: {tick}, acting agents: {len(actions)}, action counts " + \
+             f"move: {cnt_action[Move]}, attack: {cnt_action[Attack]}, " + \
+             f"sell: {cnt_action[Sell]}, use: {cnt_action[Move]}, " + \
+             f"give: {cnt_action[Give]}, buy: {cnt_action[Buy]}"
+  logging.info(info_str)
+
+  return cnt_action
+
+
+class ScriptedAgentTestConfig(nmmo.config.Small, nmmo.config.AllGameSystems):
+
+  __test__ = False
+
+  LOG_ENV = True
+
+  LOG_MILESTONES = True
+  LOG_EVENTS = False
+  LOG_VERBOSE = False
+
+  SPECIALIZE = True
+  PLAYERS = [
+    baselines.Fisher, baselines.Herbalist,
+    baselines.Prospector,baselines.Carver, baselines.Alchemist,
+    baselines.Melee, baselines.Range, baselines.Mage]
+
+
+# pylint: disable=abstract-method,duplicate-code
+class ScriptedAgentTestEnv(nmmo.Env):
+  '''
+  EnvTest step() bypasses some differential treatments for scripted agents
+  To do so, actions of scripted must be serialized using the serialize_actions function above
+  '''
+  __test__ = False
+
+  def __init__(self, config: nmmo.config.Config, seed=None):
+    super().__init__(config=config, seed=seed)
+    # this is to cache the actions generated by scripted policies
+    self.actions = {}
+
+  def reset(self, map_id=None, seed=None, options=None):
+    self.actions = {}
+    # manually resetting the EntityState, ItemState datastore tables
+    EntityState.State.table(self.realm.datastore).reset()
+    ItemState.State.table(self.realm.datastore).reset()
+    return super().reset(map_id=map_id, seed=seed, options=options)
+
+  def _compute_scripted_agent_actions(self, actions):
+    # all agent must be scripted agents when using ScriptedAgentTestEnv
+    for ent in self.realm.players.values():
+      assert isinstance(ent.agent, baselines.Scripted), 'All agent must be scripted.'
+
+    # if actions are not provided, generate actions using the scripted policy
```

Comment:
So, the current implementation is very specific to the deterministic replay test using scripted agents. 

The current action output of scripted agents seem to be different from the trained agents, so that's why the current code seems weird. I'll try to tackle this in the next iteration by looking at scripted/baselines.py.

Does this answer your question?